lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -25,16 +25,225 @@
 using namespace circt;
 using namespace firrtl;
 
+static const char seqMemAnnoClass[] =
+    "sifive.enterprise.firrtl.SeqMemInstanceMetadataAnnotation";
+static const char dutAnnoClass[] = "sifive.enterprise.firrtl.MarkDUTAnnotation";
+/// Attribute that indicates where some json files should be dumped.
+static const char metadataDirectoryAnnoClass[] =
+    "sifive.enterprise.firrtl.MetadataDirAnnotation";
+
 namespace {
 class CreateSiFiveMetadataPass
     : public CreateSiFiveMetadataBase<CreateSiFiveMetadataPass> {
   LogicalResult emitRetimeModulesMetadata();
   LogicalResult emitSitestBlackboxMetadata();
+  LogicalResult emitMemoryMetadata();
   void getDependentDialects(mlir::DialectRegistry &registry) const override;
   void runOnOperation() override;
+
+  // The set of all modules underneath the design under test module.
+  DenseSet<Operation *> dutModuleSet;
+  // The design under test module.
+  FModuleOp dutMod;
 };
 } // end anonymous namespace
 
+/// This function collects all the firrtl.mem ops and creates a verbatim op with
+/// the relevant memory attributes.
+LogicalResult CreateSiFiveMetadataPass::emitMemoryMetadata() {
+
+  // Lambda to get the number of read, write and read-write ports corresponding
+  // to a MemOp.
+  auto getNumPorts = [&](MemOp op, size_t &numReadPorts, size_t &numWritePorts,
+                         size_t &numReadWritePorts) {
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+      auto portKind = op.getPortKind(i);
+      if (portKind == MemOp::PortKind::Read)
+        ++numReadPorts;
+      else if (portKind == MemOp::PortKind::Write) {
+        ++numWritePorts;
+      } else
+        ++numReadWritePorts;
+    }
+  };
+
+  CircuitOp circuitOp = getOperation();
+  // The instance graph analysis will be required to print the hierarchy names
+  // of the memory.
+  auto instancePathCache = InstancePathCache(getAnalysis<InstanceGraph>());
+
+  auto printAttr = [&](llvm::json::OStream &J, Attribute &verifValAttr,
```

Comment:
nit but please change `J` to `j` or better yet `json` to follow casing conventions